strawberry/channels/handlers/http_handler.py

```diff
@@ -0,0 +1,238 @@
+"""GraphQLHTTPHandler
+
+A consumer to provide a graphql endpoint, and optionally graphiql.
+"""
+
+import dataclasses
+import json
+from pathlib import Path
+from typing import Any, Optional
+from urllib.parse import parse_qs
+
+from channels.db import database_sync_to_async
+from channels.generic.http import AsyncHttpConsumer
+from strawberry.channels.context import StrawberryChannelsContext
+from strawberry.exceptions import MissingQueryError
+from strawberry.http import (
+    GraphQLHTTPResponse,
+    GraphQLRequestData,
+    parse_query_params,
+    parse_request_data,
+    process_result,
+)
+from strawberry.schema import BaseSchema
+from strawberry.schema.exceptions import InvalidOperationTypeError
+from strawberry.types import ExecutionResult
+from strawberry.types.graphql import OperationType
+
+from .base import ChannelsConsumer
+
+
+class MethodNotAllowed(Exception):
+    ...
+
+
+class ExecutionError(Exception):
+    ...
+
+
+@dataclasses.dataclass
+class Result:
+    response: bytes
+    status: int = 200
+    content_type: str = "application/json"
+
+
+class GraphQLHTTPConsumer(ChannelsConsumer, AsyncHttpConsumer):
+    """A consumer to provide a view for GraphQL over HTTP.
+
+    To use this, place it in your ProtocolTypeRouter for your channels project:
+
+    ```
+    from strawberry.channels import GraphQLHttpRouter
+    from channels.routing import ProtocolTypeRouter
+    from django.core.asgi import get_asgi_application
+
+    application = ProtocolTypeRouter({
+        "http": URLRouter([
+            re_path("^graphql", GraphQLHTTPRouter(schema=schema)),
+            re_path("^", get_asgi_application()),
+        ]),
+        "websocket": URLRouter([
+            re_path("^ws/graphql", GraphQLWebSocketRouter(schema=schema)),
+        ]),
+    })
+    ```
+    """
+
+    graphiql_html_file_path = (
+        Path(__file__).parent.parent.parent / "static" / "graphiql.html"
+    )
+
+    def __init__(
+        self,
+        schema: BaseSchema,
+        graphiql: bool = True,
+        allow_queries_via_get: bool = True,
+        subscriptions_enabled: bool = True,
+        **kwargs,
+    ):
+        self.schema = schema
+        self.graphiql = graphiql
+        self.allow_queries_via_get = allow_queries_via_get
+        self.subscriptions_enabled = subscriptions_enabled
+        super().__init__(**kwargs)
+
+    async def handle(self, body: bytes):
+        try:
+            if self.scope["method"] == "GET":
+                result = await self.get(body)
+            elif self.scope["method"] == "POST":
+                result = await self.post(body)
+            else:
+                raise MethodNotAllowed()
+        except MethodNotAllowed:
+            await self.send_response(
+                405,
+                b"Method not allowed",
+                headers=[b"Allow", b"GET, POST"],
```

Comment:
This results in a ValueError in `daphne/server.py` (line 240:  `for k, v in message.get("headers", []):`
```suggestion
                headers=[(b"Allow", b"GET, POST")],
```

Thanks for building this, it's been working really well for us!